### PR TITLE
feat: add profile-specific context condensing thresholds

### LINKED ISSUE #4118: Context Condensing - Suggestion to Improve
### What problem does this proposed feature solve?

I love this feature. I really find it wonderful. The one thing that would make it really perfect would be to be able to set a different Threshold per API Config. 

### Describe the proposed solution in detail

Personally, I like to have Google Gemini 2.5 Pro condense at around 50% as my Orchestrator. But if I set it to 50%, my Code mode using Sonnet 4 ends up condensing nonstop. I would set my Sonnet 4 to more like 100% or 90% if I was able to.

### Technical considerations or implementation details (optional)

_No response_

### Describe alternatives considered (if any)

_No response_

### Additional Context & Mockups

_No response_

### Proposal Checklist

- [x] I have searched existing Issues and Discussions to ensure this proposal is not a duplicate.
- [x] This proposal is for a specific, actionable change intended for implementation (not a general idea).
- [x] I understand that this proposal requires review and approval before any development work begins.

### Are you interested in implementing this feature if approved?

- [ ] Yes, I would like to contribute to implementing this feature.

### PR DESCRIPTION
**Pull request by SannidhyaSah**
### Related GitHub Issue

Related to: #4342  and Closes #4118 

### Description

This PR implements profile-specific context condensing thresholds, addressing one of the context management improvements outlined in issue #4342 (It also closes #4118) . This feature allows users to configure different automatic threshold percentages for intelligent context condensing based on their API configuration profiles. Previously, there was only a single global threshold that applied to all profiles, which was inefficient since different LLM models have varying context window sizes.

**Key implementation details:**
- Added new settings `profileSpecificThresholdsEnabled` and `profileThresholds` to the global settings schema
- Extended the sliding window algorithm to check for profile-specific thresholds before falling back to the global threshold
- Created a new UI component `ProfileThresholdManager` for managing profile-specific configurations
- Implemented support for `-1` threshold values to indicate using the default global threshold
- Added comprehensive state management and persistence for the new settings

**What reviewers should focus on:**
- The logic in `src/core/sliding-window/index.ts` for threshold resolution
- The new UI component integration in the settings panel
- State synchronization between frontend and backend via `vscode.postMessage`

### Test Procedure

**Manual Testing:**
1. Open VSCode with the extension installed
2. Navigate to Settings → Context Management
3. Enable "Profile-Specific Thresholds" checkbox
4. Select a profile from the dropdown and set a custom threshold (e.g., 75)
5. Save the configuration and verify it appears in the list
6. Test with `-1` value to ensure it falls back to global threshold
7. Start a conversation and verify the correct threshold is applied based on the active profile

**Automated Testing:**
- Run `pnpm test` to execute all test suites
- New tests cover both backend logic and frontend UI components
- Tests verify threshold resolution logic, UI state management, and edge cases

### Type of Change

- [ ] 🐛 **Bug Fix**: Non-breaking change that fixes an issue.
- [x] ✨ **New Feature**: Non-breaking change that adds functionality.
- [ ] 💥 **Breaking Change**: Fix or feature that would cause existing functionality to not work as expected.
- [ ] ♻️ **Refactor**: Code change that neither fixes a bug nor adds a feature.
- [ ] 💅 **Style**: Changes that do not affect the meaning of the code (white-space, formatting, etc.).
- [ ] 📚 **Documentation**: Updates to documentation files.
- [ ] ⚙️ **Build/CI**: Changes to the build process or CI configuration.
- [ ] 🧹 **Chore**: Other changes that don't modify `src` or test files.

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Code Quality**:
    - [x] My code adheres to the project's style guidelines.
    - [x] There are no new linting errors or warnings (`npm run lint`).
    - [x] All debug code (e.g., `console.log`) has been removed.
- [x] **Testing**:
    - [x] New and/or updated tests have been added to cover my changes.
    - [x] All tests pass locally (`npm test`).
    - [x] The application builds successfully with my changes.
- [x] **Branch Hygiene**: My branch is up-to-date (rebased) with the `main` branch.
- [x] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [ ] **Changeset**: A changeset has been created using `npm run changeset` if this PR includes user-facing changes or dependency updates.
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

**Before:** Only global threshold setting available

image

**After:** New profile-specific threshold management UI with:
- Checkbox to enable/disable profile-specific thresholds
- Dropdown to select API configuration profiles
- Input field for threshold percentage (supports -1 for default)
- Save/Remove buttons for managing configurations
- List display of configured profile thresholds

image

### Documentation Updates

- [ ] No documentation updates are required.
- [x] Yes, documentation updates are required. The settings documentation should be updated to explain the new profile-specific threshold feature and how to configure it.

### Additional Notes

This feature significantly improves the user experience for users working with multiple LLM models that have different context window sizes. The implementation maintains backward compatibility - existing users will continue to use the global threshold until they explicitly enable profile-specific thresholds.

The UI design follows the existing patterns in the settings panel and includes proper internationalization support with translation keys added to the English locale file.

This PR addresses one aspect of the context management improvements discussed in issue #4342. Additional PRs may be needed to fully address all the concerns raised in that issue.

### Get in Touch

Discord: Lews

----

> [!IMPORTANT]
> Adds profile-specific context condensing thresholds, enhancing context management flexibility and efficiency with new settings, UI components, and comprehensive testing.
> 
>   - **Behavior**:
>     - Introduces profile-specific context condensing thresholds in `sliding-window/index.ts`, allowing different thresholds per profile.
>     - Adds `profileSpecificThresholdsEnabled` and `profileThresholds` to settings schema in `global-settings.ts`.
>     - Supports `-1` as a threshold value to use the global default.
>   - **UI**:
>     - New `ProfileThresholdManager` component in `ContextManagementSettings.tsx` for managing profile-specific thresholds.
>     - Updates settings panel to include profile selection and threshold configuration.
>   - **Testing**:
>     - Extensive tests in `sliding-window.test.ts` for threshold logic and edge cases.
>     - UI tests in `ContextManagementSettings.test.tsx` to verify new component behavior.
>   - **Misc**:
>     - Updates to `ClineProvider.ts` and `Task.ts` for state management and threshold application.
>     - Adds i18n support for new UI elements in multiple language files.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 7d3619603e288cb856ad0c9ce081232d6b7ffb37. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by samhvw8** [comment]
this is cool, since each model has difference context window and effective context window, this will be crucial pr feature :D

**Comment by mrubens** [comment]
What do you think of something like this where we just migrate them to all be profile-specific: https://github.com/RooCodeInc/Roo-Code/pull/2655? Not sure we need a checkbox for this.

**Comment by SannidhyaSah** [comment]
> What do you think of something like this where we just migrate them to all be profile-specific: https://github.com/RooCodeInc/Roo-Code/pull/2655? Not sure we need a checkbox for this.

I initially considered that placement. However, I've positioned this setting alongside the other auto-condensing options for a couple of reasons: it's directly related to that function, and it becomes irrelevant if auto-condensing is disabled.
Regarding the checkbox, you're right. It's unnecessary, and I'll remove it.
Please advise on the placement decision so I can proceed. I am currently implementing the rest of the settings page anyways .

**Comment by mrubens** [comment]
Hmm, I deleted an earlier comment. I'm having trouble wrapping my head around the right user experience for this section that lets you both choose a specific profile or just use the current one, and now to also make the threshold a per-profile configuration. 🤔 

Screenshot 2025-06-08 at 7 55 52 AM

**Comment by mrubens** [comment]
I actually think the best approach might be to put a profile selector dropdown above the threshold slider in my screenshot, the same as what we have in the section below. So you're configuring "Default" by default, but can choose another.

**Comment by SannidhyaSah** [comment]
> Hmm, I deleted an earlier comment. I'm having trouble wrapping my head around the right user experience for this section that lets you both choose a specific profile or just use the current one, and now to also make the threshold a per-profile configuration. 🤔 
> 
> Screenshot 2025-06-08 at 7 55 52 AM
> 

The API configuration for context condensing needs to be accessible even when automatic triggering is off, as it's also used for the manual trigger .
To fix this, I suggest we make the API and prompt settings always visible in the 'Context' tab. We can then group the settings that are only for automatic triggering into a separate section that expands when the user enables it

**Comment by mrubens** [comment]
> The API configuration for context condensing needs to be accessible even when automatic triggering is off, as it's also used for the manual trigger . To fix this, I suggest we make the API and prompt settings always visible in the 'Context' tab. We can then group the settings that are only for automatic triggering into a separate section that expands when the user enables it

Good call

**Comment by SannidhyaSah** [comment]
image
@mrubens is this better ?

**Comment by daniel-lxs** [comment]
Hey @SannidhyaSah, Can you take a look at the conflicts?

**Comment by SannidhyaSah** [comment]
@daniel-lxs hey . done. Please review them.

**Comment by daniel-lxs** [comment]
Hey @SannidhyaSah, Am I doing something wrong here?

https://github.com/user-attachments/assets/74161107-1bac-4329-92d6-87c3685e872d

I make a change but it doesn't seem to be permanent.

Note: I tested this with and without my commits, same result.

**Comment by SannidhyaSah** [comment]
> Hey @SannidhyaSah, Am I doing something wrong here?
> 2025-06-14_12-15-04.mp4
> 
> I make a change but it doesn't seem to be permanent.
> 
> Note: I tested this with and without my commits, same result.

Definitely not. It was working before the merge. seem like it broke after it. Let me work on it and fix it asap

**Comment by mrubens** [comment]
Looks like some conflicts here. Let me know when they're fixed and I'll review, excited about this. Thanks!

**Review by daniel-lxs — APPROVED** [review]
Took a while but the conflicts are solved

**Comment by mrubens** [comment]
Is the test failure legit?

**Comment by daniel-lxs** [comment]
@mrubens 
The test failure was legit, just needed to update the data-testid. 
It looks good now.

File: src/core/condense/index.ts [hunk 1]
```diff
@@ -8,6 +8,8 @@ import { ApiMessage } from "../task-persistence/apiMessages"
 import { maybeRemoveImageBlocks } from "../../api/transform/image-cleaning"
 
 export const N_MESSAGES_TO_KEEP = 3
+export const MIN_CONDENSE_THRESHOLD = 10 // Minimum percentage of context window to trigger condensing
```
Comment (mrubens) [inline src/core/condense/index.ts]: We changed to 5% in main - I think it makes sense to match that.

```suggestion
export const MIN_CONDENSE_THRESHOLD = 5 // Minimum percentage of context window to trigger condensing
```
